Modules/posixmodule.c

```diff
@@ -9143,8 +9148,72 @@ os_pwritev_impl(PyObject *module, int fd, PyObject *buffers, Py_off_t offset,
 }
 #endif /* HAVE_PWRITEV */
 
+#ifdef HAVE_COPY_FILE_RANGE
+/*[clinic input]
+
+os.copy_file_range
+    src: int
+        Source file descriptor.
+    dst: int
+        Destination file descriptor.
+    count: int
+        Number of bytes to copy.
+    offset_src: object = None
+        Starting offset in src.
+    offset_dst: object = None
+        Starting offset in dst.
+
+Copy count bytes from one file descriptor to another.
+
+If offset_src is None, then src is read from the current position;
+respectively for offset_dst.
+[clinic start generated code]*/
+
+static PyObject *
+os_copy_file_range_impl(PyObject *module, int src, int dst, int count,
+                        PyObject *offset_src, PyObject *offset_dst)
+/*[clinic end generated code: output=9f105e6cef25d2d1 input=25f27461ce4ddc9d]*/
+{
+    off_t offset_src_val, offset_dst_val;
+    off_t *p_offset_src = NULL;
+    off_t *p_offset_dst = NULL;
+    Py_ssize_t ret;
+    int async_err = 0;
+    /* The flags argument is provided to allow
+     * for future extensions and currently must be to 0. */
+    int flags = 0;
+
+    if (offset_src != Py_None) {
+        if (!Py_off_t_converter(offset_src, &offset_src_val)) {
+            return NULL;
+        }
+        p_offset_src = &offset_src_val;
+    }
+
+    if (offset_dst != Py_None) {
+        if (!Py_off_t_converter(offset_dst, &offset_dst_val)) {
+            return NULL;
+        }
+        p_offset_dst = &offset_dst_val;
+    }
+
+    do {
+        Py_BEGIN_ALLOW_THREADS
+        _Py_BEGIN_SUPPRESS_IPH
+        /* The call using syscall instead of copy_file_range is to maintain compatibility
+         * with glibc<2.27 */
+        ret = syscall(__NR_copy_file_range, src, p_offset_src, dst, p_offset_dst, count, flags);
```

Comment:
This contradicts your documentation and _configure.ac_ changes.

Reply:
I was (am) a bit confused about the documentation because it mentions that

> The copy_file_range() system call first appeared in Linux 4.5, but
> glibc 2.27 provides a user-space emulation when it is not available.

but in the example it mentions that

>  /* On versions of glibc before 2.27, we must invoke copy_file_range()
>          using syscall(2) */

So I am not sure which versions of glibc provide `copy_file_range` as a **directly** emulated version and which ones provides it though `syscall(2)` and which ones do not provide anything. For now and for making the patch easier to review I am going to use `copy_file_range` directly and we can relax the `#ifdef` if we want to provide compatibility with older versions.

Reply:
If we would want to support older versions, we can modify the macro as:

```
#ifdef HAVE_COPY_FILE_RANGE || defined(__GLIBC__) && ((__GLIBC__ == 2 && __GLIBC_MINOR__ == 27) || (__GLIBC__ > 2))
```

and the call as

```
#ifdef HAVE_COPY_FILE_RANGE 
ret = copy_file_range(src, p_offset_src, dst, p_offset_dst, size_count, flags);
#else
ret = syscall(__NR_copy_file_range, src, p_offset_src, dst, p_offset_dst, count, flags);
#endif
```